Libraries/LibDebug/Dwarf/DwarfInfo.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020, Itamar S. <itamar8910@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+#include "CompilationUnit.h"
+#include "DwarfTypes.h"
+#include <AK/BufferStream.h>
+#include <AK/ByteBuffer.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/RefCounted.h>
+#include <AK/String.h>
+#include <Libraries/LibELF/Loader.h>
+
+namespace Dwarf {
+
+class DwarfInfo : public RefCounted<DwarfInfo> {
+public:
+    static NonnullRefPtr<DwarfInfo> create(NonnullRefPtr<const ELF::Loader> elf) { return adopt(*new DwarfInfo(elf)); }
```

Comment:
You can `move()` in situations like this to generate slightly tighter code:
```suggestion
    static NonnullRefPtr<DwarfInfo> create(NonnullRefPtr<const ELF::Loader> elf) { return adopt(*new DwarfInfo(move(elf))); }
```

Reply:
What you actually need is `&&` on the argument type:

```cpp
static NonnullRefPtr<DwarfInfo> create(NonnullRefPtr<const ELF::Loader>&& elf) { return adopt(*new DwarfInfo(elf)); }
```

which communicates to the caller that they're losing ownership of the loader if they pass it into the function. You can then use or not use `move()` in the body.

Reply:
Yes, but in this case the caller doesn't want to give it up, so using `NonnullRefPtr&&` would force the caller into the awkward position of having to synthesize a temporary copy.

Reply:
If it doesn't want to give it up, then how is it ok to take it form under them by casting the reference to `&&` (which is what `move()` does)?

If it *does* want to give it up, it should use `move()` itself.

Reply:
The argument is a `NonnullRefPtr`, a copy made by the caller. `move()` here only yanks the argument value, not whatever the caller had on his side. :)

Reply:
Ah, it's a `NonnullRefPtr`, I was thinking it's an `NonnullOwnPtr` for some reason :facepalm: